text/2540-formal-function-parameter-attributes.md

```diff
@@ -0,0 +1,242 @@
+- Feature Name: formal_function_param_attrs
+- Start Date: 2018-10-14
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to allow attributes in formal function parameter position.
+
+# Motivation
+[motivation]: #motivation
+
+Having attributes on formal function parameters allows for certain different use cases.
+
+## Example: Handling of unused parameter
+
+In today's Rust it is possible to prefix the name of an identifier to silence the compiler about it being unused.
+With attributes in formal function parameter position we could have an attribute like `#[unused]` that explicitely states this for a given parameter.
+
+```rust
+fn foo(#[unused] bar: u32) -> bool;
+```
+
+Instead of
+
+```rust
+fn foo(_bar: u32) -> bool
+```
+
+This would better reflect the explicit nature of Rust compared to the underscore prefix as of today.
+
+## Example: Low-level code
+
+For raw pointers that are oftentimes used when operating with C code one could provide the compiler with additional information about the set of parameters.
+You could for example mirror C's restrict keyword or even be more explicit by stating what pointer argument might overlap.
+
+```rust
+fn foo(#[overlaps_with(in_b) in_a: *const u8, #[overlaps_with(in_a)] in_b: *const u8, #[restrict] out: *mut u8);
+```
+
+Which might state that the pointers `in_a` and `in_b` might overlap but `out` is non overlapping.
+Please note that I am *not* proposing to actually add this to the language!
+
+## Example: Procedural Macros
+
+Also procedural macros could greatly benefit from having their own defined custom attributes on formal parameters.
```

Comment:
I think the RFC would benefit from having a list of examples.
It doesn't have to be very detailed, but something more to go on than mere assertion would be nice.

Reply:
You are right, however, I am not sure what I could actually show as an example for a proc macro in an RFC. Like the entire implementation? Or just an example of a usage of one?

Reply:
Each example could just be a few words about what the example proc macro would do and how it would use the example attributes. If you are having a hard time with that, reach out to folks who have written many such macros. For example, rocket.rs might be a good place to ask.

You could also consider tools that interpret macros that don't do anything.